### ISSUE TITLE
HDF5 input structure

### ISSUE THREAD

**Issue by rpachaly** [FLO-2DSoftware/qgis-flo-2d-plugin#1158]

**Comment by rpachaly** [comment]
@FLO-2DKaren @FLO-2DNoemi 

I've been working on this simultaneously with the training. Here is the current state:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/f50ea36d-8c97-4f78-9226-2d988b2fc4f2)

I'm exporting everything from QGIS into these folders and subfolders with attributes to explain the meaning of each table. This can be easily changed, but I tried to keep the exact same name and format as the *.DAT files. See the TOLER as example:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/7ab22c20-5834-40f4-9a29-85e16ac31451)

I think we should change some of the data structure on the HDF5 format, but my idea here is to start simple, close to what we already have on the *.DAT files, and then start changing the file's structure as we go.

There are still some data that needs to be exported, I'll be working on those. I'm really acquainted with the code right now. 

Furthermore, I'm using compression and it is working REALLY well. I tested it on the Subdomain SD 3 project, the *.DAT files have around 187 mb and it goes down to 28 mb.

Let me know what you think and what can be improved.  Don't need to hurry, this can be discussed after the training.

**Comment by rpachaly** [comment]
RAIN.DAT

![Image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/dced2b8a-93de-4026-a816-122490c032f3)

**Comment by FLO-2DNoemi** [comment]
Robson, this looks good to me.

Let's do something similar with all the required data, then send it to me
so I can read it with FORTRAN.

Remember to change the type of data from string to the corresponding type.

**Comment by rpachaly** [comment]
CONT and TOLER -> They are ordered based on the Data Input Manual sequence that they appear on the CONT.DAT and TOLER.DAT files

![Image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/0112ec97-7855-421f-a583-19ec2e369c5e)

Since we're using floats now, I set the null values to -9999.

**Comment by rpachaly** [comment]
GRID group


![Image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/40df1a1d-fd11-4707-8396-c6e024478f3d)

Maybe group X and Y?

**Comment by FLO-2DNoemi** [comment]
This looks good to me.

**Comment by FLO-2DNoemi** [comment]
Yes, we can group X and Y.

**Comment by rpachaly** [comment]
LEVEE -> This is the one that I did more changes on the structure, I created a specific dataset for levee failure and other for the levee data. However, I could not create a situation on the plugin where the fragility curve is used. @FLO-2DKaren, is the fragility curve for levees available on the plugin?

![Image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/489af7e3-0857-4bdb-ab46-9b2d3ce707e9)

**Comment by rpachaly** [comment]
@FLO-2DNoemi Here is an hdf5 file with the already implemented datasets. When we have integers and floats on the same dataset, I'm setting the whole dataset to float. That's why in some cases the grid number is a float and in others it is an integer. 

[test.zip](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/files/14963469/test.zip)

**Comment by FLO-2DNoemi** [comment]
Robson,

Received.  I will look into it and let you know my comments.

Thanks
N

**Comment by FLO-2DKaren** [comment]
Hi Robson,

There are 3 components for the plugin that are not finished because no one uses them.
Evaporation
Streets
Levee Fragility Curves

I'm not certain what to do about this.  Noemi might have some ideas.

@FLO-2DNoemi @rpachaly 

When you do levees, you might use a method like this where the 1 - 8 is crest elevation.
It would be a float table.
If the value is full that is the cutoff direction.  If the value is NULL, it is no levee in that direction.

grid  1  2  3  4  5  6  7  8

I didn't look at your hdf5.  So if you already did something else or better, cool.

**Comment by rpachaly** [comment]
SCS method:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/e2b16375-da0f-4d2f-82c3-9d263d69e18e)

Obs: INFIL_METHOD is always present if different than 0
Obs: INFIL_SCS_CELLS only appears if the infiltration was calculated for each cell. Otherwise, only INFIL_SCS_GLOBAL appears.

**Comment by rpachaly** [comment]
Perfect! I'm working on the infiltration.

**Comment by rpachaly** [comment]
HORTON:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/1085cf42-3ea5-481d-b283-525c65836204)

Obs: INFIL_HORTON_CELLS only appears if the infiltration was calculated for each cell. Otherwise, only INFIL_HORTON_GLOBAL appears.

**Comment by rpachaly** [comment]
GREEN-AMPT:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/3d333e31-7bf4-4999-af33-c58a9faa2116)

Obs: INFIL_GA_CELLS only appears if the infiltration was calculated for each cell. Otherwise, only INFIL_GA_GLOBAL appears.

**Comment by rpachaly** [comment]
Channel Infiltration:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/e15233e6-bf3b-448e-9c29-cf7dc39f41eb)

These datasets are created only if INFCHAN = 1 on the INFIL_GA_GLOBAL.

Obs: I couldn't create the INFIL_CHAN_ELEMS using the plugin. I had to change the 'Channel Elements' table .

**Comment by rpachaly** [comment]
ARF:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/1354c609-943d-4549-a0a5-adb19d238325)

**Comment by FLO-2DNoemi** [comment]
I do not remember getting a question or working on a project in the last
few years that includes any of these.

I think we do not need to add 1D streets. I do not see the need to model
streets with the 1D component in the future.

Evaporation is important to calculate for dispersion of pollutants so we
should eventually add the feature to the plugin.  Let's add this and the
fragility curves to the github as an issue. Let's finish the first HDF5 and
key developments for plugin 1.0, then we can add these two
additional features.

**Comment by rpachaly** [comment]
For the OUTFLOW I did a video to explain how I organized the data. It is different, but I think it is better. Check the video and let me know what you think. 

https://flo-2d.sharefile.com/d-sf1440a3701af431a99a4b54d96a7516d

**Comment by rpachaly** [comment]
INFLOW

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/f27e9374-79a9-42e6-981e-77a7f46d046d)

The small caps are the ones that are different from the INFLOW.DAT

**Comment by FLO-2DKaren** [comment]
@FLO-2DNoemi @rpachaly 

Hi Robson and Noemi,  There is an error when you don't turn on the n value switch.  FLOPRO won't run.  So maybe we should just have that be on default.

**Comment by rpachaly** [comment]
FPXSEC

This one was a little tricky. I believe that this is the best format. The only issue is that, in order to keep all values as integers, I had to add the -9999 to the columns where no grid is present.

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/4e1cf47f-e1e9-491b-8190-b3f839d77165)

**Comment by rpachaly** [comment]
TOLSPATIAL

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/f0646701-0e8f-4ffa-bbf9-9d58766e794e)

**Comment by rpachaly** [comment]
SHALLOWN_SPATIAL

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/76035ef4-d28a-4ba4-b545-2b9fa6e2844b)

**Comment by rpachaly** [comment]
MULT and SIMPLE_MULT

Hi @FLO-2DKaren and @FLO-2DNoemi, I was trying to export these files and I couldn't create them on the plugin. I was checking the db trigger and it is not coded yet.

**Comment by rpachaly** [comment]
> MULT and SIMPLE_MULT
> 
> Hi @FLO-2DKaren and @FLO-2DNoemi, I was trying to export these files and I couldn't create them on the plugin. I was checking the db trigger and it is not coded yet.

This was a bug related to my geopackage. Everything is working.

**Comment by rpachaly** [comment]
MULT and SIMPLE_MULT

I don't know what is the problem exactly, but sometimes adding the Multiple Channels and the Simple Multiple Channels works fine and sometimes it does not trigger the mult_cells/simple_mult_cells tables. However, the export to hdf5 is working. 

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/d3a6cf7f-6c66-4270-921e-5dd96a448619)

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/53f0f5b1-b1f1-4adf-88ab-f8b1e7450877)

**Comment by FLO-2DJJ** [comment]
I tested all the following situations without any problem:
- Whenever a **Multiple Channel Area** is created in the User Layers (_mult_areas_ table) a trigger is fired to create **Multiple Channel Cell** features (_mult_cells_ table), one feature for each cell the area intersects with the grid layer.

- Similarly, a trigger is fired when a **Multiple Channel Line** (_mult_lines_ table) is added, creating **Multiple Channel Cells** (_mult_cells_ table) from the intersection of the line with the cells of the grid layer.

- Adding **Simple Mult. Channel Lines** (in _simple_mult_lines_ table) triggers the  creation of **Simple Mult. Chann. Cells** features (in _simple_mult_cells_ table).

- Deleting any of the above features of the User Layers causes the deletion of the cells intersected either by the area or the line.

**Comment by rpachaly** [comment]
CHAN, CHANBANK and XSEC

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/10cdb8c2-2379-4bcd-be2a-6aba16afba5b)

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/1434bf51-f2a6-47b7-a93b-7e7a92fb33ff)

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/0acf6c4d-be80-430f-935c-2e0fea105253)

**Comment by FLO-2DJJ** [comment]
Hi Robson,
Thank you for the video. I see the problem there.
> Here is a video of the issue related to mult and simple_mult;
> mult.mp4
> 
> The geopackages: [mult-bug-gpkgs.zip](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/files/15092716/mult-bug-gpkgs.zip)

I suspect the triggers, for some reason, are disabled in SelfHelpKit_v1.0.0.gpkg. 
I enabled them with **self.gutils.enable_geom_triggers()** and the problem was solved.

If that's the case, not only the multiple channel triggers were disabled but all the ones defined in the _"trigger_control"_ table.

I included a **enable_geom_triggers()** in **import_selected_components2** where it was missing (It will be merged into master in my next pull request from a branch I'm working for Storage Units).

There may be some others. I'm looking at them.
Could you please check the use of **disable_geom_triggers** in the code you have been working with to see if there is any potential exception that could leave the triggers disabled? They look ok to me but I may miss the potential exceptions. Thanks.

**Comment by rpachaly** [comment]
I agree that the problem is related to the triggers, but I couldn't find where they are being disabled. 

I will check for any disable_geom_triggers on the code that I'm working on. 

Thanks for the fix JJ!

**Comment by FLO-2DNoemi** [comment]
Hi Robson,

This seems fine. I agree that it's the best format, especially considering
that FP cross-sections have relatively few cells, resulting in a low number
of no data entries.

Noemi

**Comment by FLO-2DNoemi** [comment]
Hi Robson,

I reviewed this and it looks good. I do not see any issue. I will let you
know if something arises once I read the HDF5 file.

Thanks,
N

**Comment by rpachaly** [comment]
Check this out @FLO-2DNoemi

https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/39889306/495250d3-6762-4ace-8757-e99ace5112aa

**Comment by rpachaly** [comment]
To retrieve the INP in python: 

```
with h5py.File(hdf5_file_path, 'r') as hdf:
    swmm_input_contents = hdf['SWMM_INP'][()]
    swmm_input_contents = swmm_input_contents.decode('utf-8')

    with open(output_inp_file_path, 'w') as output_file:
        output_file.write(swmm_input_contents)
```

**Comment by FLO-2DNoemi** [comment]
Hi Robson,

I like the idea of keeping the INI and INP files in the HDF5 as you
showed in the video.

The FLOPRO will still read it from INP for this upcoming release.

Noemi

**Comment by FLO-2DNoemi** [comment]
Robson,

Please try loading the INP file,  the one created from the FLO-2D HDF5 Data
file with the additional line at the end of the INP file, into EPA  SWMM
GUI and see if it works.

Thanks,

Noemi

**Comment by rpachaly** [comment]
> QGIS H5PY module updated with QGIS Version 3.36.
I don't think we will have too many problems with h5py module update or an hdf5 version update because they seem to be more compatible these days.

The h5py version that I'm using have a DeprecationWarning. I hope they had fixed it. 

> Please try loading the INP file,  the one created from the FLO-2D HDF5 Data
file with the additional line at the end of the INP file, into EPA  SWMM
GUI and see if it works.

It works. When I retrieve the INP file using python (h5py), the two files are identical.

**Comment by FLO-2DKaren** [comment]
Hi Robson,

I see you merged the code so I can test this this weekend.

It has been my experience that those hard returns are added when the file is loaded. 
I'll check it when I'm doing testing and storm drain documentation this weekend.

**Comment by rpachaly** [comment]
Yes! Everything that is commented here is implemented. Still need to export:

- dam break or breach files
- mud files
- tailing files
- building collapse

**Comment by rpachaly** [comment]
See https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/issues/997#issuecomment-2171122814

**Comment by rpachaly** [comment]
Mudflow and Sediment Transport format

![Image](https://github.com/user-attachments/assets/00039689-7f7b-4d40-8e62-2fbd9354b62e)
